### DISCUSSION TITLE
WinUI3 - CredentialPicker

### DISCUSSION THREAD

**Discussion by Balkoth** [microsoft/WindowsAppSDK#3147] [Q&A] [answered]
Is there something like this available in WinUI3? `CredentialPicker.PickAsync(...)` just throws an exception...

**Comment by castorix** [comment, marked as answer]
This test seems to work for me (Windows 10 21H1 (french), Windows app SDK 1.1.0) :

 ```
                uint nSize = 260;
                StringBuilder sbComputerName = new StringBuilder((int)nSize);
                GetComputerName(sbComputerName, ref nSize);
                var options = new CredentialPickerOptions()
                {
                    TargetName = sbComputerName.ToString(),
                    Caption = "Title",
                    Message = "This is a test",
                    CredentialSaveOption = CredentialSaveOption.Selected,
                    AuthenticationProtocol = AuthenticationProtocol.Basic,
                    AlwaysDisplayDialog = true
                };
                CredentialPickerResults results = await CredentialPicker.PickAsync(options);
```

with 

```
        [DllImport("Kernel32.dll", SetLastError = true, CharSet = CharSet.Unicode)]
        public static extern bool GetComputerName(StringBuilder lpBuffer, ref uint nSize);
```

I get :

![CredentialPicker](https://user-images.githubusercontent.com/22345506/201848748-a9810ff4-4519-40bc-bb77-3e16f0408618.jpg)

**Reply by Balkoth** [reply]
Thanks, i could get it to show, but upon closer inspection this "Helper"-class seems like a big security risk to me. The password, if using basic authentication, is returned as a string, so hangs basically around in memory until the application is closed.

My task is the following, maybe anyone can help:

- Open the dialog with a preselected user, not allowing to change the user
- Pass the returned credentials to LogonUser to verify the credentials are valid
- If invalid: Reshow the dialog with ERROR_LOGON_FAILURE as ErrorCode
- If valid: Call a PowerShell API constructing a PSCredential where a SecureString is expected for the password

I can do all this with `CredUIPromptForWindowsCredentials`, `CredPackAuthenticationBuffer`, `CredUnPackAuthenticationBuffer` and `CredUIParseUserName`, but this is a tedious and also badly underdocumented approach.
